File: web/static/diagram_click.js
/* Robust mermaid node → chapter (#32). data-id / click row, not fragile textContent.
   Also honors #3 diagram-nodes blob URLs when present. */
(function (global) {
  function extractNodeId() {
    for (let i = 0; i < arguments.length; i += 1) {
      const text = String(arguments[i] || "");
      const isolated = text.match(/(?:^|[\s#_.-])(A\d+)(?:$|[\s#_.-])/);
      if (isolated) return isolated[1];
      if (/^A\d+$/.test(text.trim())) return text.trim();
    }
    return null;
  }

  function parseJsonScript(id, root) {
    const el = root.getElementById ? root.getElementById(id) : null;
    if (!el) return [];
    try { return JSON.parse(el.textContent || "[]"); } catch { return []; }
  }

  function resolveDiagramHref(node, chapters, diagramNodes) {
    const dataId = node.getAttribute("data-id") || node.getAttribute("data-node-id") || "";
    const elementId = node.id || "";
    const clickFile = node.getAttribute("data-filename") || node.getAttribute("data-click") || "";
    const nodeId = extractNodeId(dataId, elementId);
    const nodes = diagramNodes || [];
    if (nodeId) {
      const dn = nodes.find((item) => item.id === nodeId);
      if (dn && (dn.blob || dn.href || dn.chapter)) return dn.blob || dn.href || dn.chapter;
      const hit = chapters.find((ch) => ch.node_id === nodeId);
      if (hit && hit.href) return hit.href;
    }
    if (clickFile) {
      const name = clickFile.split("/").pop();
      const byFile = chapters.find((ch) => ch.filename === name);
      if (byFile && byFile.href) return byFile.href;
    }
    const label = (node.getAttribute("data-title") || "").trim();
    if (label) {
      const dn = nodes.find((item) => String(item.title || "").trim() === label);
      if (dn && (dn.blob || dn.href || dn.chapter)) return dn.blob || dn.href || dn.chapter;
      const exact = chapters.find((ch) => String(ch.title || "").trim() === label);
      if (exact && exact.href) return exact.href;
    }
    return null;
  }

  function bindMermaidChapterClicks(root) {
    const scope = root || document;
    const raw = scope.getElementById ? scope.getElementById("chapter-nav") : null;
    if (!raw) return;
    let chapters = [];
    try { chapters = JSON.parse(raw.textContent || "[]"); } catch { return; }
    const diagramNodes = parseJsonScript("diagram-nodes", scope);
    const byId = {};
    (diagramNodes || []).forEach((item) => {
      if (item.id) {
        byId[item.id] = {
          href: item.blob || item.href || item.chapter || "",
          blob: item.blob || null,
          chapter: item.chapter || item.href || "",
          files: item.files || [],
        };
      }
    });
    scope.querySelectorAll(".mermaid .node").forEach((node) => {
      const nodeId = extractNodeId(node.getAttribute("data-id") || "", node.id || "");
      const rec = (nodeId && byId[nodeId]) || null;
      const href = resolveDiagramHref(node, chapters, diagramNodes);
      const blob = rec && rec.blob;
      const chapter = (rec && (rec.chapter || rec.href)) || href;
      const target = blob || chapter || href;
      if (!target) return;
      node.style.cursor = "pointer";
      node.setAttribute("tabindex", "0");
      if (rec && rec.files && rec.files.length) node.setAttribute("data-files", rec.files.join(","));
      if (blob) node.setAttribute("data-blob-href", blob);
      if (chapter) node.setAttribute("data-chapter-href", chapter);
      const go = () => {
        if (blob && /^https?:\/\//.test(blob)) {
          window.open(blob, "_blank", "noopener,noreferrer");
          return;
        }
        if (chapter) window.location.href = chapter;
      };
      node.addEventListener("click", (ev) => { ev.preventDefault(); ev.stopPropagation(); go(); });
      node.addEventListener("keydown", (ev) => {
        if (ev.key === "Enter" || ev.key === " ") { ev.preventDefault(); go(); }
      });
    });
  }

  global.qsDiagramClick = { extractNodeId, resolveDiagramHref, bindMermaidChapterClicks };
})(typeof window !== "undefined" ? window : globalThis);

File: utils/diagram_click.py
"""Robust diagram node → chapter routing (#32). Prefer data-id / click row."""

from __future__ import annotations

import re
from pathlib import Path

CLICK_RE = re.compile(r'^\s*click\s+(A\d+)\s+"([^"]+)"\s*$', re.M)
NODE_ID_RE = re.compile(r"(A\d+)")


def parse_mermaid_clicks(index_text: str) -> dict[str, str]:
    """Map mermaid node id (A0) → chapter filename from `click A0 "01_x.md"`."""
    out: dict[str, str] = {}
    for match in CLICK_RE.finditer(index_text or ""):
        out[match.group(1)] = Path(match.group(2).split("#", 1)[0]).name
    return out


def attach_node_ids(chapters: list[dict], index_text: str = "") -> list[dict]:
    clicks = parse_mermaid_clicks(index_text)
    by_file = {Path(fn).name: nid for nid, fn in clicks.items()}
    content = [ch for ch in chapters if ch.get("filename") not in {None, "index.md", "README.md"}]
    attached = []
    seq = 0
    for item in chapters:
        row = dict(item)
        filename = row.get("filename") or ""
        if filename in {"index.md", "README.md"}:
            row["node_id"] = None
        elif filename in by_file:
            row["node_id"] = by_file[filename]
        else:
            row["node_id"] = f"A{seq}"
            seq += 1
        attached.append(row)
    if not clicks:
        seq = 0
        for row in attached:
            if row.get("filename") not in {None, "index.md", "README.md"}:
                row["node_id"] = f"A{seq}"
                seq += 1
    return attached


def extract_node_id(*candidates: str) -> str | None:
    for raw in candidates:
        text = str(raw or "")
        if not text:
            continue
        # Prefer isolated A<digits> (data-id / flowchart-A0-123 / A0)
        match = re.search(r"(?:^|[\s#_.-])(A\d+)(?:$|[\s#_.-])", text)
        if match:
            return match.group(1)
        match = NODE_ID_RE.search(text)
        if match and text.strip() == match.group(1):
            return match.group(1)
    return None


def resolve_diagram_href(
    *,
    data_id: str = "",
    element_id: str = "",
    click_file: str = "",
    text_content: str = "",
    chapters: list[dict],
) -> str | None:
    """Match node → chapter href. data-id / click filename beat fragile textContent."""
    by_node = {ch.get("node_id"): ch for ch in chapters if ch.get("node_id")}
    by_file = {ch.get("filename"): ch for ch in chapters if ch.get("filename")}
    node_id = extract_node_id(data_id, element_id)
    if node_id and node_id in by_node:
        return by_node[node_id].get("href")
    fname = Path(click_file).name if click_file else ""
    if fname in by_file:
        return by_file[fname].get("href")
    label = (text_content or "").strip()
    if not label:
        return None
    # Exact title only — do not substring-match mixed SVG textContent.
    for ch in chapters:
        if ch.get("filename") in {None, "index.md"}:
            continue
        if (ch.get("title") or "").strip() == label:
            return ch.get("href")
    return None
